pulser-core/pulser/operators/hamiltonian.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2024 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A class to define the Pulser Hamiltonian."""
+from __future__ import annotations
+
+from pulser.operators.operator import TimeOperator
+from pulser import Sequence
+from pulser.register.base_register import BaseRegister
+from pulser.devices._device_datacls import BaseDevice
+from pulser.sampler.samples import SequenceSamples
+
+
+class Hamiltonian(TimeOperator):
+
+    @classmethod
+    def from_samples(
```

Comment:
I could use some guidance here.
I have just figured that with the new operators, we can build the Hamiltonian without Qutip. We could then take out the Hamiltonian class out of pulser-simulation such that it can be used by the other frameworks ?
In pulser-simulation, I would put a "QutipHamiltonian" that would convert the Hamiltonian into a QobjEvo. 
However, this might go a bit beyond the scope of this PR and I think it should be left as an issue after this PR is merged
What do you think @HGSilveri @a-quelle ?
Could also mention @lvignoli whom I discussed this possibility with.

Reply:
What is the need for this? I'm not familiar with pulser simulation, but I suspect it's a bit specific to there. I'm currently implementing something very similar to the `Backend` ABC, and I have a run method like
```
        coeff = 0.001  # Omega and delta are given in rad/ms, dt in ns
        dt = mps_config.dt
        omega_delta = _extract_omega_delta(sequence, dt, mps_config.with_modulation)
        emuct_register = get_qubit_positions(sequence.register)

        evolve_state = MPS(
            len(sequence.register.qubits),
            truncate=False,
            precision=mps_config.precision,
            max_bond_dim=mps_config.max_bond_dim,
            num_devices_to_use=mps_config.num_devices_to_use,
        ) 

        for step in range(omega_delta.shape[1]):
            t = step * dt
            mpo_t = make_H(
                emuct_register,
                omega_delta[0, step, :],
                omega_delta[1, step, :],
                sequence.device.interaction_coeff,
            )
            evolve_tdvp(-coeff * dt * 1j, evolve_state, mpo_t, mps_config.max_krylov_dim)

        return result
```
so there is no real reason to explicitly store a time series anywhere. If the user wants to measure an operator, the can create it from the `Operator` ABC that you created below.

Reply:
The idea with this Hamiltonian(TimeOperator) is that it would avoid you making _extract_omega_delta, get_qubit_positions, and make_H, you could just do Hamiltonian.from_sequence(sequence, sampling_times), that would make a list of mpo_t (as Operator) and time steps, and then you could convert these Operators into mpo_t (as matrix product states, with the converter you have defined).
That could enable us to share the same definition of the Pulser Hamiltonian ? I think this could be useful for other emulators as well.

Reply:
For code efficiency reasons, I would avoid that code path even if it were available. Having an abstract representation for operators and states is unavoidable if you want a backend independent interface, but within the backend, everything goes. Constructing the Hamiltonians by going through the abstract format that we're now defining would be at least an order of magnitude slower than the bit fiddling I'm doing in `make_H`, as well as creating problems with differentiability due to `svd` operations in the general algorithm `MPO` addition algorithm that I took some pains to avoid in `make_H`.

Reply:
I appreciate that you're trying to be helpful, but that is not your job. Rather it's my job to be helpful by removing the burden what logic is needed to actually efficiently simulate a pulse in a given backend from the pulser team.
For historical reasons, it doesn't quite work that way yet, but there is no time like the present to try and improve things.

Reply:
I also think this would severely complicate things for marginal benefit. If the Hamiltonian was a single operator I could see it, but the time-dependence seems like a deal breaker to me

Reply:
Thanks for your feedback @a-quelle and @HGSilveri, let's forget about this feature then.